hashmap/src

```diff
@@ -0,0 +1,248 @@
+            return Err(MeasurementStreamError::UnexpectedEndOfData.into());
+        }
+
+        self.buffer.push(b'}');
+        println!("vector {:#?}", std::str::from_utf8(&self.buffer));
+        Ok(())
+    }
+
+    fn timestamp(&mut self, value: String) -> Result<(), Self::Error> {
+        if self.is_within_group {
+            return Err(MeasurementStreamError::UnexpectedTimestamp.into());
+        }
+
+        if self.needs_separator {
+            self.buffer.push(b',');
+        }
+        self.buffer
+            .write_fmt(format_args!("\"time\":\"{}\"", value))?;
+        self.needs_separator = true;
+        Ok(())
+    }
+
+    fn measurement(&mut self, name: &str, value: f64) -> Result<(), Self::Error> {
+        if self.needs_separator {
+            self.buffer.push(b',');
+        }
+        self.buffer
+            .write_fmt(format_args!("\"{}\":{}", name, value))?;
+        self.needs_separator = true;
+        Ok(())
+    }
+
+    fn start_group(&mut self, group: &str) -> Result<(), Self::Error> {
+        if self.is_within_group {
+            return Err(MeasurementStreamError::UnexpectedStartOfGroup.into());
+        }
+
+        if self.needs_separator {
+            self.buffer.push(b',');
+        }
+        self.buffer.write_fmt(format_args!("\"{}\":{{", group))?;
+        self.needs_separator = false;
+        Ok(())
+    }
+
+    fn end_group(&mut self) -> Result<(), Self::Error> {
+        if !self.is_within_group {
+            return Err(MeasurementStreamError::UnexpectedEndOfGroup.into());
+        }
+
+        self.buffer.push(b'}');
+        self.needs_separator = true;
+        Ok(())
+    }
+}
+
+fn main() {
+    println!("Hello, world!");
+    let mut tej_producer = ThinEdgeJsonMap::new();
+    tej_producer.timestamp("4-20-2020".into()).unwrap();
+    tej_producer.measurement(None, "temperature", 25.0).unwrap();
+    tej_producer
+        .measurement(Some("location"), "alti", 2100.4)
+        .unwrap();
+    tej_producer
+        .measurement(Some("location"), "longi", 2100.4)
+        .unwrap();
+    tej_producer
+        .measurement(Some("location"), "lati", 2100.4)
+        .unwrap();
+    tej_producer
+        .measurement(Some("location"), "alti", 2100.5)
+        .unwrap();
+
+    println!("values--->{:#?}", tej_producer);
+
+    //Serialize the TEJ to u8 bytes
+    let mut t_serializer = ThinEdgeJsonSerializer::new();
+    t_serializer.start();
+    t_serializer.timestamp(tej_producer.timestamp);
+    for (key, value) in tej_producer.values.iter() {
+        match value {
+            Measurement::Single(sv) => {
+                t_serializer.measurement(key,*sv);
+            }
+            Measurement::Multi(m) => {
+                t_serializer.start_group(key);
+                for (key, value) in m.iter() {
+                    t_serializer.measurement(key, *value);
+                }
+                t_serializer.end_group();
+            }
+        }
+    }
+    t_serializer.end();
```

Comment:
This must be implemented as a method of the producer here a `ThinEdgeJsonMap` struct.

Something along the lines:

```
let bytes = tej_producer.build(ThinEdgeJsonSerializer::new())
``` 

* What is the better name for this function? `build`, `visit`, `fold` ... This is a fold but build might be more appropriate.
* The type would be `fn build<C, E, D>(&self, collector: C) -> Result<D, E>  where C: MeasurementConsumer<Error = E, Data = D>`.

Reply:
Agree, I also have the same plan, to implement the serializer as part of the ThinEdgeJsonMap.

Reply:
Also let bytes =  tej_producer.serialize_tej() is enough right. The serialize_tej function creates the serializer and pushes out a vector of bytes.